src/Build/BackEnd/Components/RequestBuilder/AssemblyLoadsTracker.cs

```diff
@@ -0,0 +1,102 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+//
+
+using System;
+using System.Collections.Concurrent;
+using Microsoft.Build.BackEnd.Logging;
+using Microsoft.Build.Framework;
+
+namespace Microsoft.Build.BackEnd.Components.RequestBuilder
+{
+    internal class AssemblyLoadsTracker : MarshalByRefObject, IDisposable
+    {
+        private static readonly ConcurrentDictionary<AppDomain, AssemblyLoadsTracker> s_instances = new();
+        private readonly LoggingContext? _loggingContext;
+        private readonly LoggingService? _loggingService;
+        private readonly AssemblyLoadingContext _context;
+        private readonly AppDomain _appDomain;
+
+        private AssemblyLoadsTracker(
+            LoggingContext? loggingContext,
+            LoggingService? loggingService,
+            AssemblyLoadingContext context,
+            AppDomain appDomain)
+        {
+            _loggingContext = loggingContext;
+            _loggingService = loggingService;
+            _context = context;
+            _appDomain = appDomain;
+        }
+
+        public static IDisposable StartTracking(
+            LoggingContext loggingContext,
+            AssemblyLoadingContext context,
+            AppDomain? appDomain = null)
+        {
+            return StartTracking(loggingContext, null, context, appDomain);
+        }
+
+        public static IDisposable StartTracking(
+            LoggingService loggingService,
+            AssemblyLoadingContext context,
+            AppDomain? appDomain = null)
+        {
+            return StartTracking(null, loggingService, context, appDomain);
+        }
+
+        public static void StopTracking(AppDomain appDomain)
+        {
+            if (s_instances.TryRemove(appDomain, out AssemblyLoadsTracker? tracker))
+            {
+                tracker.StopTracking();
+            }
+        }
+
+        public void Dispose()
+        {
+            StopTracking();
+        }
+
+        private static IDisposable StartTracking(
+            LoggingContext? loggingContext,
+            LoggingService? loggingService,
+            AssemblyLoadingContext context,
+            AppDomain? appDomain = null)
+        {
```

Comment:
Can you put this behind a 17.6 changewave opt-out, so if we discover some functional problem we can easily disable it?